Add support for ClickHouse

kmichel-aiven opened the pull request:
Add support for ClickHouse database backup and restore.

kmichel-aiven committed:
plugins: add clickhouse support

This does not include partial restore, all nodes must
be restored together on an empty ClickHouse cluster.

kmichel-aiven committed:
clickhouse: add tests

kmichel-aiven committed:
coordinator: don't require a truthy value for all steps

This avoids the awkward Union[bool, something] return types and replaces
the return False with a StepFailedError to represent errors.

kmichel-aiven committed:
clickhouse: add dependencies

kmichel-aiven committed:
clickhouse: tolerate failure during a restore operation

Two main things were improved to allow for a successful restore after
a failed one :
- Previously created database are dropped before creating them again,
  this also takes care of cleaning partially restored tables.
- Previously created entities are ignored instead of failing when
  trying to create them again.

These two steps are now tested for this behavior.

The main integration test was also modified to test the behavior of
an entire restore after a failure on any step, this greatly increases
the tests duration, the entire suite takes almost 2 minutes now.

dmitry-potepalov commented on tests/unit/coordinator/plugins/clickhouse/test_dependencies.py:
```diff
@@ -0,0 +1,95 @@
+    assert tables_sorted_by_dependencies([t1]) == [t1]
+
+
+def test_access_entities_sorted_by_dependencies():
```
maybe add a test for a dangling access entity dependency?

dmitry-potepalov reviewed (commented):
Overall good job, I enjoyed reading this). Some minor issues/questions in comments.

kmichel-aiven committed:
tests: fix type hints

We don't run mypy on tests/ during CI but we probably should.

kmichel-aiven committed:
tests: fix client timeout tests

- Refine exception type
- Remove timeout set twice
- Test no server and slow server timeout
- Don't kill the shared Clickhouse service

kmichel-aiven committed:
clickhouse: make SYNC REPLICA timeout configurable

We can't expect data synchronization to be finished in 10 seconds
on a realistically sized cluster. Make it configurable and increase
default value to 3600 seconds. (Other queries stay at 10 seconds).

kmichel-aiven committed:
clickhouse: move cleanup step earlier

This reduces the gap between fetching
table schemas and freezing the tables.

kmichel-aiven committed:
clickhouse: merge databases and tables retrieval

Databases created after the RetrieveDatabasesStep were correctly
ignored when fetching tables, but dropped databases would be still
present and restored without their attached tables, restoring a state
that never existed originally. Merging the two steps and their query
solves this problem and... reduces the number of steps.

kmichel-aiven commented on astacus/coordinator/plugins/clickhouse/client.py:
```diff
@@ -0,0 +1,103 @@
+    ClickHouse client using the HTTP(S) protocol, the port provided in the url
+    should match the `http_port` or `https_port` server option.
+
+    The url should not contain a database suffix, the client forces all query to
```
Outdated comment.

The initial code was using an URL, but when testing the corner cases of client side in a project using Astacus,
I discovered too many traps (unescaped password, ipv6 brackets in hostnames...), so I swiched to separate fields.

The database customization wouldn't be complicated, but not really desirable, the `system` database is hardcoded
is ClickHouse and is a fake database for metadata that is guaranteed to be there, we don't really want to connect
using something else.

kmichel-aiven commented on astacus/coordinator/plugins/clickhouse/plugin.py:
```diff
@@ -0,0 +1,90 @@
+
+class ClickHousePlugin(CoordinatorPlugin):
+    zookeeper: ZooKeeperConfiguration = ZooKeeperConfiguration()
+    clickhouse: ClickHouseConfiguration = ClickHouseConfiguration()
```
A bit confusing yes, but a `CoordinatorPlugin` is a pydantic model and directly created from the config file, it would be strange to have "zookeeper_config" in the configuration file itself.

It can seem strange too that, in a sense, the plugin is its own config. This removes the need to write create functions to create a plugin after we have created the typed config, and it forces a pattern where the plugin `__init__` do not have side effect and can always be created without worry during tests.

kmichel-aiven commented on astacus/coordinator/plugins/clickhouse/plugin.py:
```diff
@@ -0,0 +1,90 @@
+            RetrieveReplicatedDatabasesStep(clients=clickhouse_clients),
+            RetrieveTablesStep(clients=clickhouse_clients),
+            # Cleanup previously forgotten frozen parts (this won't remove frozen parts from deleted tables)
+            RemoveFrozenTablesStep(freeze_name=self.freeze_name),
```
Good catch, It was dependent but it's not anymore (hence the comment, that is now wrong).

It's much better to move it before the `Retrieve*` :+1:

kmichel-aiven commented on tests/integration/coordinator/plugins/clickhouse/test_client.py:
```diff
@@ -0,0 +1,50 @@
+async def test_client_execute_bounded_failure_time(clickhouse: Service):
+    client = get_clickhouse_client(clickhouse, timeout=1.0)
+    client.timeout = 1
+    clickhouse.process.kill()
```
Good catch, I just got lucky because it was the last test of the module...

I still need the kill test though, the HTTP library behaves quite differently between
a slow connection opening and and slow query (I added the slow query test though).

(pausing the server with SIGSTOP / SIGCONT does not work with ClickHouse.)

fingon reviewed (dismissed):
Disclaimer: Not full review, as I just read the generic code parts. If you wish for full review from me, ping me explicitly on Slack (this is .. painfully large.. PR, and I don't know that much about Clickhouse so if you have some subject matter expert in your team all the better) :-)

kmichel-aiven committed:
tests: add type hints

kmichel-aiven committed:
don't rethrow WaitResultError as StepFailedError

dmitry-potepalov reviewed (approved):
LGTM
